### PR TITLE
Use hmac for digest nonces

### PR DESCRIPTION
**Pull request by ceeram**
Use hmac for digest nonces

### CODE REVIEW

File: src/Auth/DigestAuthenticate.php [hunk 1]
```diff
@@ -250,7 +251,7 @@ public function loginHeaders(ServerRequest $request)
     protected function generateNonce()
     {
         $expiryTime = microtime(true) + $this->getConfig('nonceLifetime');
-        $signatureValue = md5($expiryTime . ':' . $this->getConfig('secret'));
+        $signatureValue = hash_hmac('sha256', $expiryTime . ':' . $this->getConfig('secret'), Security::getSalt());
```
Comment (markstory) [inline src/Auth/DigestAuthenticate.php]: Don't you have to change the validation of the nonce as well?

Comment (ceeram) [reply src/Auth/DigestAuthenticate.php]: yes it should, will take care of that

Comment (markstory) [reply src/Auth/DigestAuthenticate.php]: There is also a generateNonce method in the test case that is still using md5() as well.

Comment (ceeram) [reply src/Auth/DigestAuthenticate.php]: updated both

File: tests/TestCase/Auth/DigestAuthenticateTest.php [hunk 2]
```diff
@@ -515,7 +515,7 @@ protected function generateNonce($secret = null, $expires = 300, $time = null)
         $secret = $secret ?: Configure::read('Security.salt');
         $time = $time ?: microtime(true);
         $expiryTime = $time + $expires;
-        $signatureValue = md5($expiryTime . ':' . $secret);
+        $signatureValue = hash_hmac('sha1', $expiryTime . ':' . $secret, $secret);
```
Comment (timalive) [inline tests/TestCase/Auth/DigestAuthenticateTest.php]: Is this just me, or shouldn't this be sha256?  seems to be causing some trouble for me.

File: src/Auth/DigestAuthenticate.php [hunk 3]
```diff
@@ -276,7 +277,8 @@ protected function validNonce($nonce)
         if ($expires < microtime(true)) {
             return false;
         }
+        $check = hash_hmac('sha1', $expires . ':' . $this->getConfig('secret'), $this->getConfig('secret'));
```
Comment (timalive) [inline src/Auth/DigestAuthenticate.php]: Is this just me, or shouldn't this be sha256? seems to be causing some trouble for me.

File: tests/TestCase/Auth/DigestAuthenticateTest.php (hunk 2 above)
Comment (markstory) [reply tests/TestCase/Auth/DigestAuthenticateTest.php]: Are you suggesting sha256 for the stronger algorithm?

Comment (timalive) [reply tests/TestCase/Auth/DigestAuthenticateTest.php]: no, due to inconsistency with the validNonce function

Comment (markstory) [reply tests/TestCase/Auth/DigestAuthenticateTest.php]: I totally missed that :blush:. I'll get it fixed.

Comment (timalive) [reply tests/TestCase/Auth/DigestAuthenticateTest.php]: great! Thanks. Had me pulling my hair out.

**Pull request merged**